src/libgit2/config.c

```diff
@@ -593,36 +593,67 @@ static int get_backend_for_use(git_config_backend **out,
 	git_config *cfg, const char *name, backend_use use)
 {
 	size_t i;
+	size_t len;
 	backend_internal *backend;
+	int error = 0;
+	git_config_entry *entry = NULL;
+	char *key = NULL;
 
 	*out = NULL;
 
-	if (git_vector_length(&cfg->backends) == 0) {
+	len = git_vector_length(&cfg->backends);
+	if (len == 0) {
 		git_error_set(GIT_ERROR_CONFIG,
 			"cannot %s value for '%s' when no config backends exist",
 			uses[use], name);
 		return GIT_ENOTFOUND;
 	}
 
 	git_vector_foreach(&cfg->backends, i, backend) {
-		if (!backend->backend->readonly) {
-			*out = backend->backend;
-			return 0;
+		if (backend->backend->readonly)
+			continue;
+
+		/* git-config doesn't update worktree-level config
```

Comment:
This comment is interesting — isn't this true of all `git config` entry levels? IOW, doesn't `git config` always operate on the local configuration unless otherwise specified?

I actually think that this exposes an oddity in our whole ordering. Our configuration levels appear to be trying to support a scenario where we have a system of ordering so that higher priority backends can override the data in lower priority backends. (eg, if I have `foo.bar` in system, and `foo.bar` in local, then the `foo.bar` in local is used.)

But we also want a system that's relatively flexible so that library users can plug in their own backends. (In other words, we _don't_ want to just always default to the local backend, which is what we _would_ do in the absence of any pluggability.)

The wrinkle here is that you're introducing a new backend that should take priority over reads but should _not_ take priority over writes. I think that we should consider doing two things:

1. For the sort of generic, unopinionated configuration system, we should have two priority ordered lists — one for read priority, and one for write priority
2. When a repository is created by default, it should set the read priority in the order that we've specified here. The write priority should only have a single option, which is the local configuration. (By default, calling set without a specific entry level should probably just always go there.)

This lets users have control over the ordering, but sets a default that aligns with git.

Does this make sense? Am I entirely off base? Am I overthinking things?

Reply:
I don't think you're over-thinking things. I agree that the different behavior for reading/writing is not something that's nicely explicit in the current API.

This feels very related to https://github.com/libgit2/libgit2/pull/6202#discussion_r1501780259. I will have to noodle on these points more (I've got some household chores to do in the meantime), but I suspect my other reply might be a start toward a resolution here.